hudi-client/src/main/java/org/apache/hudi/HoodieWriteClient.java

```diff
@@ -367,20 +370,30 @@ public static SparkConf registerClasses(SparkConf conf) {
     }
   }
 
+  private BulkInsertMapFunction<T> getBulkInsertMapFunction(
+      boolean isSorted, String commitTime, HoodieWriteConfig config, HoodieTable<T> hoodieTable,
+      List<String> fileIDPrefixes) {
+    if (isSorted) {
+      return new BulkInsertMapFunctionForSortedRecords(
+          commitTime, config, hoodieTable, fileIDPrefixes);
+    }
+    return new BulkInsertMapFunctionForNonSortedRecords(
+        commitTime, config, hoodieTable, fileIDPrefixes);
+  }
+
   private JavaRDD<WriteStatus> bulkInsertInternal(JavaRDD<HoodieRecord<T>> dedupedRecords, String commitTime,
       HoodieTable<T> table, Option<UserDefinedBulkInsertPartitioner> bulkInsertPartitioner) {
```

Comment:
I thought `UserDefinedBulkInsertPartitioner` is user-facing so I didn't touch the interface.  Looking at the source code again, I find that `UserDefinedBulkInsertPartitioner` is only used here.  Then we can just keep a single one as `BulkInsertInternalPartitioner `.

BTW, is this interface (`UserDefinedBulkInsertPartitioner`) intended for user to implement a custom partitioner?  I don't seem to see a config to pass in a custom implementation.  wondering if we should provide such flexibility.

Reply:
This was added by @ovj and uber/marmaray project passes in an implementation using the RDD api

Reply:
@yihua Yes, we do need to keep this functionality, at Uber as @vinothchandar pointed out, uses this. We directly pass an implementation rather than a config.